API Calls failing with exception - Entity has already been requested. Entity cannot be requested multiple times

We have a set of API endpoints exposed in our helidon application and after upgrading to Helidon 4, it fails with the above exception

## Environment Details
* Helidon Version: 4.0.4 (Upgrading from 3.2.2)
* Helidon MP
* JDK version: 21
* OS: Mac OS
* Docker version (if applicable): NA

----------

## Problem Description
We are trying to upgrade our Helidon application from v3 to v4. After the upgrade, requests to the API endpoints are failing with the exception "Entity has already been requested. Entity cannot be requested multiple times" (refer attached stack trace). The exception occurs before the logic of the endpoint comes into picture and it seems to come from the Helidon server itself. 

Following is the full stack trace.

```
Internal server error
io.helidon.http.RequestException: Entity has already been requested. Entity cannot be requested multiple times
	at io.helidon.http.RequestException$Builder.build(RequestException.java:139) ~[helidon-http-4.0.4.jar:4.0.4]
	at io.helidon.webserver.http.ErrorHandlers.unhandledError(ErrorHandlers.java:202) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at io.helidon.webserver.http.ErrorHandlers.lambda$handleError$1(ErrorHandlers.java:182) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at java.util.Optional.ifPresentOrElse(Optional.java:198) ~[?:?]
	at io.helidon.webserver.http.ErrorHandlers.handleError(ErrorHandlers.java:181) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at io.helidon.webserver.http.ErrorHandlers.runWithErrorHandling(ErrorHandlers.java:118) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at io.helidon.webserver.http.Filters$FilterChainImpl.proceed(Filters.java:121) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at io.helidon.webserver.observe.metrics.MetricsFeature.lambda$configureVendorMetrics$2(MetricsFeature.java:90) ~[helidon-webserver-observe-metrics-4.0.4.jar:4.0.4]
	at io.helidon.webserver.http.Filters$FilterChainImpl.proceed(Filters.java:119) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at io.helidon.webserver.security.SecurityContextFilter.filter(SecurityContextFilter.java:88) ~[helidon-webserver-security-4.0.4.jar:4.0.4]
	at io.helidon.webserver.http.Filters$FilterChainImpl.proceed(Filters.java:119) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at io.helidon.common.context.Contexts.runInContext(Contexts.java:117) ~[helidon-common-context-4.0.4.jar:4.0.4]
	at io.helidon.webserver.observe.tracing.TracingObserver$TracingFilter.filter(TracingObserver.java:247) ~[helidon-webserver-observe-tracing-4.0.4.jar:4.0.4]
	at io.helidon.webserver.http.Filters$FilterChainImpl.proceed(Filters.java:119) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at io.helidon.common.context.Contexts.runInContext(Contexts.java:117) ~[helidon-common-context-4.0.4.jar:4.0.4]
	at io.helidon.webserver.context.ContextRoutingFeature.filter(ContextRoutingFeature.java:50) ~[helidon-webserver-context-4.0.4.jar:4.0.4]
	at io.helidon.webserver.http.Filters$FilterChainImpl.proceed(Filters.java:119) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at io.helidon.webserver.http.Filters.executeFilters(Filters.java:87) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at io.helidon.webserver.http.Filters.lambda$filter$0(Filters.java:83) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at io.helidon.webserver.http.ErrorHandlers.runWithErrorHandling(ErrorHandlers.java:75) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at io.helidon.webserver.http.Filters.filter(Filters.java:83) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at io.helidon.webserver.http.HttpRouting.route(HttpRouting.java:109) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at io.helidon.webserver.http1.Http1Connection.route(Http1Connection.java:421) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at io.helidon.webserver.http1.Http1Connection.handle(Http1Connection.java:194) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at io.helidon.webserver.ConnectionHandler.run(ConnectionHandler.java:165) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at io.helidon.common.task.InterruptableTask.call(InterruptableTask.java:47) ~[helidon-common-task-4.0.4.jar:4.0.4]
	at io.helidon.webserver.ThreadPerTaskExecutor$ThreadBoundFuture.run(ThreadPerTaskExecutor.java:239) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at java.lang.VirtualThread.run(VirtualThread.java:309) ~[?:?]
Caused by: java.lang.IllegalStateException: Entity has already been requested. Entity cannot be requested multiple times
	at io.helidon.http.media.ReadableEntityBase.inputStream(ReadableEntityBase.java:87) ~[helidon-http-media-4.0.4.jar:4.0.4]
	at io.helidon.webserver.http.ServerRequestEntity.inputStream(ServerRequestEntity.java:97) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at io.helidon.microprofile.server.JaxRsService.doHandle(JaxRsService.java:207) ~[helidon-microprofile-server-4.0.4.jar:4.0.4]
	at io.helidon.microprofile.server.JaxRsService.lambda$handle$2(JaxRsService.java:185) ~[helidon-microprofile-server-4.0.4.jar:4.0.4]
	at io.helidon.common.context.Contexts.runInContext(Contexts.java:117) ~[helidon-common-context-4.0.4.jar:4.0.4]
	at io.helidon.microprofile.server.JaxRsService.handle(JaxRsService.java:185) ~[helidon-microprofile-server-4.0.4.jar:4.0.4]
	at io.helidon.webserver.http.HttpRouting$RoutingExecutor.doRoute(HttpRouting.java:668) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at io.helidon.webserver.http.HttpRouting$RoutingExecutor.call(HttpRouting.java:627) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at io.helidon.webserver.http.HttpRouting$RoutingExecutor.call(HttpRouting.java:605) ~[helidon-webserver-4.0.4.jar:4.0.4]
	at io.helidon.webserver.http.ErrorHandlers.runWithErrorHandling(ErrorHandlers.java:75) ~[helidon-webserver-4.0.4.jar:4.0.4]
	... 22 more
```

Since we have a bunch of teams owning different API endpoints, we have consolidated them into different Applications. We have a Main function that adds these applications and starts the server as shown in the code-block below.

I have found that the issue does not affect the endpoints of the application that gets added to the server first. For the rest of the applications, the requests are met with the mentioned exception. I have compared the different applications and the corresponding resource classes and did not find any major differences between them.

```
public static void startServer() {
        log.info("Building Helidon Server");
        Server.builder()
                .addApplication(App1.class)
                .addApplication(App2.class)
                .addApplication(App3.class)
                .addApplication(App4.class)
                .build()
                .start();
        log.info("Helidon Server started successfully");
    }
```

## Steps to reproduce
- Try to create a custom startup with multiple applications registered. 
- Try to hit the endpoints to see if the same exception happens

This issue is triggered when there is more than one instance of `io.helidon.microprofile.server.JaxRsService` in the routing chain, each instance associated with a different JaxRs application.

If the request contains an entity payload and the first service does not match because the request is for a different JaxRs application, then the next service in the chain fails as the first one has already invoked `req.content().inputStream()`.

E.g.

```java
public class Main {
    public static void main(String[] args) {
        Server.builder()
                .addApplication(App1.class)
                .addApplication(App2.class)
                .build()
                .start();
    }

    @ApplicationPath("/")
    @ApplicationScoped
    public static class App1 extends Application {

        @Override
        public Set<Class<?>> getClasses() {
            return Set.of(Resource.class);
        }

        @Path("/app1")
        public static class Resource {

            @GET
            public String get() {
                return "OK";
            }
        }
    }

    @ApplicationPath("/")
    @ApplicationScoped
    public static class App2 extends Application {

        @Override
        public Set<Class<?>> getClasses() {
            return Set.of(Resource.class);
        }

        @Path("/app2")
        public static class Resource {

            @GET
            public String get() {
                return "OK";
            }
        }
    }
}
```

```shell
curl -d "{}" http://localhost:8080/app2
```

Server logs:
```
io.helidon.http.RequestException: Entity has already been requested. Entity cannot be requested multiple times
	at io.helidon.http.RequestException$Builder.build(RequestException.java:139)
	at io.helidon.webserver.http.ErrorHandlers.unhandledError(ErrorHandlers.java:202)
	at io.helidon.webserver.http.ErrorHandlers.lambda$handleError$1(ErrorHandlers.java:182)
	at java.base/java.util.Optional.ifPresentOrElse(Optional.java:198)
	at io.helidon.webserver.http.ErrorHandlers.handleError(ErrorHandlers.java:181)
	at io.helidon.webserver.http.ErrorHandlers.runWithErrorHandling(ErrorHandlers.java:118)
	at io.helidon.webserver.http.Filters$FilterChainImpl.proceed(Filters.java:121)
	at io.helidon.webserver.observe.metrics.MetricsFeature.lambda$configureVendorMetrics$2(MetricsFeature.java:90)
	at io.helidon.webserver.http.Filters$FilterChainImpl.proceed(Filters.java:119)
	at io.helidon.common.context.Contexts.runInContext(Contexts.java:117)
	at io.helidon.webserver.context.ContextRoutingFeature.filter(ContextRoutingFeature.java:50)
	at io.helidon.webserver.http.Filters$FilterChainImpl.proceed(Filters.java:119)
	at io.helidon.webserver.http.Filters.executeFilters(Filters.java:87)
	at io.helidon.webserver.http.Filters.lambda$filter$0(Filters.java:83)
	at io.helidon.webserver.http.ErrorHandlers.runWithErrorHandling(ErrorHandlers.java:75)
	at io.helidon.webserver.http.Filters.filter(Filters.java:83)
	at io.helidon.webserver.http.HttpRouting.route(HttpRouting.java:109)
	at io.helidon.webserver.http1.Http1Connection.route(Http1Connection.java:421)
	at io.helidon.webserver.http1.Http1Connection.handle(Http1Connection.java:194)
	at io.helidon.webserver.ConnectionHandler.run(ConnectionHandler.java:165)
	at io.helidon.common.task.InterruptableTask.call(InterruptableTask.java:47)
	at io.helidon.webserver.ThreadPerTaskExecutor$ThreadBoundFuture.run(ThreadPerTaskExecutor.java:239)
	at java.base/java.lang.VirtualThread.run(VirtualThread.java:309)
Caused by: java.lang.IllegalStateException: Entity has already been requested. Entity cannot be requested multiple times
	at io.helidon.http.media.ReadableEntityBase.inputStream(ReadableEntityBase.java:87)
	at io.helidon.webserver.http.ServerRequestEntity.inputStream(ServerRequestEntity.java:97)
	at io.helidon.microprofile.server.JaxRsService.doHandle(JaxRsService.java:208)
	at io.helidon.microprofile.server.JaxRsService.lambda$handle$2(JaxRsService.java:185)
	at io.helidon.common.context.Contexts.runInContext(Contexts.java:117)
	at io.helidon.microprofile.server.JaxRsService.handle(JaxRsService.java:185)
	at io.helidon.webserver.http.HttpRouting$RoutingExecutor.doRoute(HttpRouting.java:668)
	at io.helidon.webserver.http.HttpRouting$RoutingExecutor.call(HttpRouting.java:627)
	at io.helidon.webserver.http.HttpRouting$RoutingExecutor.call(HttpRouting.java:605)
	at io.helidon.webserver.http.ErrorHandlers.runWithErrorHandling(ErrorHandlers.java:75)
	... 17 more
```

---

When Helidon discovers the classes, if there is no application it aggregates all resources into a single synthetic application, in this case you won't run into this issue.

However, if there is more than one application, then the issue still exist.
E.g. replace the main method above with:

```java
io.helidon.Main.main(args);
```

Note that if the application classes are not annotated with `@ApplicationScope` you will have a single synthetic application, potentially wrongly configured (see #8502) ; in this case you won't run into this issue.

--- a/microprofile/server/src/main/java/io/helidon/microprofile/server/JaxRsService.java
+++ b/microprofile/server/src/main/java/io/helidon/microprofile/server/JaxRsService.java
@@ -205,7 +205,7 @@ class JaxRsService implements HttpService {
 
         JaxRsResponseWriter writer = new JaxRsResponseWriter(res);
         requestContext.setWriter(writer);
-        requestContext.setEntityStream(req.content().inputStream());
+        requestContext.setEntityStream(new LazyInputStream(req));
         requestContext.setProperty("io.helidon.jaxrs.remote-host", req.remotePeer().host());
         requestContext.setProperty("io.helidon.jaxrs.remote-port", req.remotePeer().port());
         requestContext.setRequestScopedInitializer(ij -> {

--- a/microprofile/server/src/main/java/io/helidon/microprofile/server/LazyInputStream.java
+++ b/microprofile/server/src/main/java/io/helidon/microprofile/server/LazyInputStream.java
@@ -1,0 +1,107 @@
+/*
+ * Copyright (c) 2024 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.microprofile.server;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+import io.helidon.common.LazyValue;
+import io.helidon.webserver.http.ServerRequest;
+
+/*
+ Only obtains the input stream from request when used.
+ This is to work around the case where we have a Jersey application before other routing for requests with entity.
+ Before this fix, we would request the input stream, and the next component would fail with input stream already requested.
+ */
+class LazyInputStream extends InputStream {
+    private final LazyValue<InputStream> delegate;
+
+    LazyInputStream(ServerRequest req) {
+        delegate = LazyValue.create(() -> req.content().inputStream());
+    }
+
+    @Override
+    public int read() throws IOException {
+        return delegate.get().read();
+    }
+
+    @Override
+    public int read(byte[] b) throws IOException {
+        return delegate.get().read(b);
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        return delegate.get().read(b, off, len);
+    }
+
+    @Override
+    public byte[] readAllBytes() throws IOException {
+        return delegate.get().readAllBytes();
+    }
+
+    @Override
+    public byte[] readNBytes(int len) throws IOException {
+        return delegate.get().readNBytes(len);
+    }
+
+    @Override
+    public int readNBytes(byte[] b, int off, int len) throws IOException {
+        return delegate.get().readNBytes(b, off, len);
+    }
+
+    @Override
+    public long skip(long n) throws IOException {
+        return delegate.get().skip(n);
+    }
+
+    @Override
+    public void skipNBytes(long n) throws IOException {
+        delegate.get().skipNBytes(n);
+    }
+
+    @Override
+    public int available() throws IOException {
+        return delegate.get().available();
+    }
+
+    @Override
+    public void close() throws IOException {
+        delegate.get().close();
+    }
+
+    @Override
+    public void mark(int readlimit) {
+        delegate.get().mark(readlimit);
+    }
+
+    @Override
+    public void reset() throws IOException {
+        delegate.get().reset();
+    }
+
+    @Override
+    public boolean markSupported() {
+        return delegate.get().markSupported();
+    }
+
+    @Override
+    public long transferTo(OutputStream out) throws IOException {
+        return delegate.get().transferTo(out);
+    }
+}

--- a/tests/integration/mp-gh-8349/pom.xml
+++ b/tests/integration/mp-gh-8349/pom.xml
@@ -1,0 +1,63 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+    Copyright (c) 2024 Oracle and/or its affiliates.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <groupId>io.helidon.tests.integration</groupId>
+        <artifactId>helidon-tests-integration</artifactId>
+        <version>4.0.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>helidon-tests-integration-mp-gh-8349</artifactId>
+    <name>Helidon Tests Integration MP GH 8349</name>
+    <description>Reproducer for Github issue #8349 - Entity already requested after Jersey returns 404</description>
+
+    <dependencies>
+        <dependency>
+            <groupId>io.helidon.microprofile.server</groupId>
+            <artifactId>helidon-microprofile-server</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.helidon.microprofile.metrics</groupId>
+            <artifactId>helidon-microprofile-metrics</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.helidon.logging</groupId>
+            <artifactId>helidon-logging-jul</artifactId>
+            <scope>runtime</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.hamcrest</groupId>
+            <artifactId>hamcrest-all</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.helidon.microprofile.testing</groupId>
+            <artifactId>helidon-microprofile-testing-junit5</artifactId>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+</project>

--- a/tests/integration/mp-gh-8349/src/main/java/io/helidon/tests/integration/gh8349/Gh8349App1.java
+++ b/tests/integration/mp-gh-8349/src/main/java/io/helidon/tests/integration/gh8349/Gh8349App1.java
@@ -1,0 +1,48 @@
+/*
+ * Copyright (c) 2024 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.tests.integration.gh8349;
+
+import java.util.Set;
+
+import io.helidon.common.Weight;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.ws.rs.ApplicationPath;
+import jakarta.ws.rs.POST;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.core.Application;
+import jakarta.ws.rs.core.MediaType;
+
+@ApplicationScoped
+@ApplicationPath("/")
+@Weight(100)
+public class Gh8349App1 extends Application {
+    @Override
+    public Set<Class<?>> getClasses() {
+        return Set.of(Gh8349Resource1.class);
+    }
+
+    @Path("/greet1")
+    public static class Gh8349Resource1 {
+        @POST
+        @Produces(MediaType.TEXT_PLAIN)
+        public String testPost(String entity) {
+            return "Hello World 1!";
+        }
+    }
+}

--- a/tests/integration/mp-gh-8349/src/main/java/io/helidon/tests/integration/gh8349/Gh8349App2.java
+++ b/tests/integration/mp-gh-8349/src/main/java/io/helidon/tests/integration/gh8349/Gh8349App2.java
@@ -1,0 +1,48 @@
+/*
+ * Copyright (c) 2024 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.tests.integration.gh8349;
+
+import java.util.Set;
+
+import io.helidon.common.Weight;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.ws.rs.ApplicationPath;
+import jakarta.ws.rs.POST;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.core.Application;
+import jakarta.ws.rs.core.MediaType;
+
+@ApplicationScoped
+@ApplicationPath("/")
+@Weight(50)
+public class Gh8349App2 extends Application {
+    @Override
+    public Set<Class<?>> getClasses() {
+        return Set.of(Gh8349Resource2.class);
+    }
+
+    @Path("/greet2")
+    public static class Gh8349Resource2 {
+        @POST
+        @Produces(MediaType.TEXT_PLAIN)
+        public String testPost(String entity) {
+            return "Hello World 2!";
+        }
+    }
+}

--- a/tests/integration/mp-gh-8349/src/main/resources/META-INF/beans.xml
+++ b/tests/integration/mp-gh-8349/src/main/resources/META-INF/beans.xml
@@ -1,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+    Copyright (c) 2024 Oracle and/or its affiliates.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+<beans xmlns="https://jakarta.ee/xml/ns/jakartaee"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="https://jakarta.ee/xml/ns/jakartaee
+                           https://jakarta.ee/xml/ns/jakartaee/beans_4_0.xsd"
+        version="4.0"
+        bean-discovery-mode="annotated">
+</beans>

--- a/tests/integration/mp-gh-8349/src/main/resources/logging.properties
+++ b/tests/integration/mp-gh-8349/src/main/resources/logging.properties
@@ -1,0 +1,22 @@
+#
+# Copyright (c) 2024 Oracle and/or its affiliates.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+handlers=io.helidon.logging.jul.HelidonConsoleHandler
+java.util.logging.SimpleFormatter.format=%1$tY.%1$tm.%1$td %1$tH:%1$tM:%1$tS %4$s %3$s !thread!: %5$s%6$s%n
+
+.level=WARNING
+
+io.helidon.level=INFO

--- a/tests/integration/mp-gh-8349/src/test/java/io/helidon/tests/integration/gh8349/Gh8349Test.java
+++ b/tests/integration/mp-gh-8349/src/test/java/io/helidon/tests/integration/gh8349/Gh8349Test.java
@@ -1,0 +1,64 @@
+/*
+ * Copyright (c) 2024 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.tests.integration.gh8349;
+
+import io.helidon.http.Status;
+import io.helidon.microprofile.testing.junit5.HelidonTest;
+
+import jakarta.inject.Inject;
+import jakarta.ws.rs.client.Entity;
+import jakarta.ws.rs.client.WebTarget;
+import jakarta.ws.rs.core.Response;
+import org.junit.jupiter.api.Test;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+@HelidonTest
+class Gh8349Test {
+    private final WebTarget target;
+
+    @Inject
+    Gh8349Test(WebTarget target) {
+        this.target = target;
+    }
+
+    @Test
+    void testApp1() {
+        try (Response response = target.path("/greet1")
+                .request()
+                .post(Entity.text("hello"))) {
+            assertThat(response.getStatus(), is(Status.OK_200.code()));
+
+            String entity = response.readEntity(String.class);
+            assertThat(entity, is("Hello World 1!"));
+        }
+
+    }
+
+    @Test
+    void testApp2() {
+        try (Response response = target.path("/greet2")
+                .request()
+                .post(Entity.text("hello"))) {
+            assertThat(response.getStatus(), is(Status.OK_200.code()));
+
+            String entity = response.readEntity(String.class);
+            assertThat(entity, is("Hello World 2!"));
+        }
+    }
+}

--- a/tests/integration/pom.xml
+++ b/tests/integration/pom.xml
@@ -53,6 +53,7 @@
         <module>mp-gh-4123</module>
         <module>mp-gh-4654</module>
         <module>mp-gh-5328</module>
+        <module>mp-gh-8349</module>
         <module>mp-gh-8478</module>
         <module>mp-gh-8495</module>
         <module>mp-graphql</module>

